### PR TITLE
#28: Initial integration with network policies

### PR DESCRIPTION
**Pull request by glothriel**

### CODE REVIEW

File: pkg/k8s/exposer.go [hunk 1]
```diff
@@ -2,131 +2,118 @@
...
-	clientset, clientSetErr := kubernetes.NewForConfig(config)
+	return kubernetes.NewForConfig(config)
+}
+
+type k8sResourceExposer struct {
+	namespace string
+	child     listeners.Exposer
+	selectors map[string]string
+
+	managedResources []managedK8sResource
```
Comment (jzyinq) [inline pkg/k8s/exposer.go:35]: big fat +1 for that - that will be much more manageable imo later on

File: tests/fixtures.py [hunk 2]
```diff
@@ -388,64 +453,6 @@ def download(url, path):
         f.write(response.content)
 
 
-class Nginx:
```
Comment (glothriel) [inline tests/fixtures.py:391]: This was dead code

**Comment by glothriel** [comment]
The cluster for integration tests was changed to K3d, so it also implements https://github.com/glothriel/wormhole/issues/73

File: README.md [hunk 3]
```diff
@@ -100,6 +100,40 @@ wormhole.glothriel.github.com/ports=http
...
+    podSelector:
+        matchLabels:
+            application: wormhole-client-dev1
+    policyTypes:
+    - Ingress
+```
+
+Such policies allow communication from any pod in any namespace, providing, that the pod that tries to communicate has a label `wormhole.glothriel.github.com/network-policy-consumes-app` with the value of the name of the service that is exposed. The app name (unless override by `wormhole.glothriel.github.com/name=my-custom-name`) is `<service-namespace-name>-<service-name>` (for example `default-nginx`) of the service exposed from remote cluster.
+
+Effectively this means, that the permission to communicate is granted per application, not per peer. Having permission to communicate with app having given name, allows the pod to communicate with all the apps with given name, no matter the peer the app is exposed from.
```
Comment (jzyinq) [inline README.md:135]: So iuuic that means that server does not need multiple network policies to be able to reach for example, some exposed nginx on each of those clients. 

It's just needs something like:
```yaml
  template:
    metadata:
      labels:
        app: server
        wormhole.glothriel.github.com/network-policy-consumes-app: nginx-nginx
```
For clients this isn't very important.

Maybe you could rephrase a bit to emphasize that.

File: pkg/k8s/exposer.go [hunk 4]
```diff
@@ -2,131 +2,114 @@
...
 
-func (factory *k8sServiceExposer) Withdraw(app peers.App) error {
-	serviceName := fmt.Sprintf("%s-%s", app.Peer, app.Name)
-	logrus.Debugf("Deleting service %s", serviceName)
-	config, inClusterConfigErr := rest.InClusterConfig()
-	if inClusterConfigErr != nil {
-		return inClusterConfigErr
-	}
-	clientset, clientSetErr := kubernetes.NewForConfig(config)
+func (exp *k8sResourceExposer) WithdrawAll() error {
```
Comment (jzyinq) [inline pkg/k8s/exposer.go:79]: Is it used anywhere else then along tests? What's the usecase for such withdrawal? Part of graceful shutdown of whole wormhole on the specific cluster?

File: pkg/k8s/exposer.go [hunk 5]
```diff
@@ -2,131 +2,114 @@
...
+}
+
+func (exp *k8sResourceExposer) Add(app peers.App) (peers.App, error) {
+	clientset, clientSetErr := exp.clientProvider.New()
 	if clientSetErr != nil {
 		return peers.App{}, clientSetErr
 	}
-	servicesClient := clientset.CoreV1().Services(factory.namespace)
-	addedApp, childFactoryErr := factory.child.Add(app)
+	addedApp, childFactoryErr := exp.child.Add(app)
```
Comment (jzyinq) [inline pkg/k8s/exposer.go:45]: I'm not a fan of nesting exposer within exposer - shoudn't we have some other interface that will accept a list of `exposers` and work on them accordingly?

`wormHoleKeeper` ? ;)

**Review by jzyinq — APPROVED** [review]
works like intended - good job!

File: README.md [hunk 6]
```diff
@@ -100,6 +100,40 @@ wormhole.glothriel.github.com/ports=http
...
+        - port: 25001
+            protocol: TCP
+    podSelector:
+        matchLabels:
+            application: wormhole-client-dev1
+    policyTypes:
+    - Ingress
+```
+
+Such policies allow communication from any pod in any namespace, providing, that the pod that tries to communicate has a label `wormhole.glothriel.github.com/network-policy-consumes-app` with the value of the name of the service that is exposed. The app name (unless override by `wormhole.glothriel.github.com/name=my-custom-name`) is `<service-namespace-name>-<service-name>` (for example `default-nginx`) of the service exposed from remote cluster.
```
[the remark this answers is not in the document]
Comment (glothriel) [reply README.md:133]: `default-nginx` will be replaced by whatever you typed

File: README.md (hunk 3 above)
Comment (glothriel) [reply README.md:135]: Ok, i'll try to emphasize that

File: pkg/k8s/exposer.go (hunk 4 above)
Comment (glothriel) [reply pkg/k8s/exposer.go:79]: Good point, it's unused. I wanted to use it when wormhole starts to cleanup all remnants of old state. https://github.com/glothriel/wormhole/issues/76

File: pkg/k8s/exposer.go (hunk 5 above)
Comment (glothriel) [reply pkg/k8s/exposer.go:45]: I'd say it is what it is :D I didn't want to do a refactor of half the codebase. Saying that we can discuss it. Decorator is my favorite design pattern (as you probably noticed) and I may have hammer/nail problem with it.

**Pull request merged**